### PR TITLE
JIRA-XXX implementation of a FilteringManagement API using Event Sour…

### PR DESCRIPTION
**Pull request by mbaechler**
…cing

Ok, java is very verbose and ES doesn't help here ...

### CODE REVIEW

**Comment by mbaechler** [comment]
test this please

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/Rule.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        Rule rule = (Rule) o;
+        return Objects.equals(id, rule.id);
+    }
+
+    @Override
+    public int hashCode() {
```
Comment (chibenwa) [inline server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/Rule.java]: You miss final declaration & probably equalsVerifier...

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/impl/DefineRulesCommand.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    public List<Rule> getRules() {
+        return rules;
+    }
+
+    public User getUser() {
+        return user;
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
Comment (chibenwa) [inline server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/impl/DefineRulesCommand.java]: Idem

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/impl/FilteringAggregate.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    }
+
+    public List<Rule> listRules() {
+        return state.rules;
+    }
+
+    private void apply(Event event) {
+        if (event instanceof RuleSetDefined) {
+            state = state.set(((RuleSetDefined)event).getRules());
+        }
```
Comment (chibenwa) [inline server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/impl/FilteringAggregate.java]: What happens if we don't know the event?

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/impl/RuleSetDefined.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    public AggregateId getAggregateId() {
+        return aggregateId;
+    }
+
+    public ImmutableList<Rule> getRules() {
+        return rules;
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
Comment (chibenwa) [inline server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/impl/RuleSetDefined.java]: idem

File: server/data/data-jmap/src/test/java/org/apache/james/jmap/api/filtering/FilteringAggregateIdTest.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+    @Test
+    void asAggregateKeyShouldReturnAStringContainingThePrefixAndTheDomain() {
+        assertThat(new FilteringAggregateId(User.fromUsername("foo@bar.space")).asAggregateKey())
+            .isEqualTo("FilteringRule/foo@bar.space");
+    }
+
+    @Test
+    void parseShouldThrowWhenNullString() {
+        assertThatThrownBy(() -> FilteringAggregateId.parse(null)).isInstanceOf(NullPointerException.class);
```
Comment (chibenwa) [inline server/data/data-jmap/src/test/java/org/apache/james/jmap/api/filtering/FilteringAggregateIdTest.java]: Line breaks?

File: server/data/data-jmap/src/test/java/org/apache/james/jmap/api/filtering/FilteringAggregateIdTest.java [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    void parseShouldThrowWhenStringDoesntMatchPrefix() {
+        assertThatThrownBy(() -> FilteringAggregateId.parse("WrongPrefix/foo"))
+            .isInstanceOf(IllegalArgumentException.class);
+    }
+
+    @Test
+    void parseShouldThrowWhenStringDoesntHavePrefix() {
+        assertThatThrownBy(() -> FilteringAggregateId.parse("FilteringRulefoo"))
+            .isInstanceOf(IllegalArgumentException.class);
+    }
```
Comment (chibenwa) [inline server/data/data-jmap/src/test/java/org/apache/james/jmap/api/filtering/FilteringAggregateIdTest.java]: I can think of two other edge cases... 'FilteringRule' alone and 'FilteringRule/'

File: server/data/data-jmap/src/test/java/org/apache/james/jmap/api/filtering/FilteringManagementContract.java [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+import com.google.common.collect.ImmutableList;
+
+public interface FilteringManagementContract {
+
+    FilteringManagement instanciateFilteringManagement();
+
+    @Test
+    default void listingRulesForUnknownUserShouldReturnEmptyList() {
+        User user = User.fromUsername("bart@simpson.cartoon");
```
Comment (chibenwa) [inline server/data/data-jmap/src/test/java/org/apache/james/jmap/api/filtering/FilteringManagementContract.java]: Can this user be a constant?

File: server/data/data-jmap/src/test/java/org/apache/james/jmap/api/filtering/FilteringManagementContract.java [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+    @Test
+    default void definingEmptyRuleListShouldRemoveExistingRules() {
+        User user = User.fromUsername("bart@simpson.cartoon");
+        FilteringManagement testee = instanciateFilteringManagement();
+        testee.defineRulesForUser(user, ImmutableList.of(Rule.of("3"), Rule.of("2"), Rule.of("1")));
+        testee.defineRulesForUser(user, ImmutableList.of());
+        assertThat(testee.listRulesForUser(user)).isEmpty();
+    }
+
```
Comment (chibenwa) [inline server/data/data-jmap/src/test/java/org/apache/james/jmap/api/filtering/FilteringManagementContract.java]: I can not see the `set null rules` edge case

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/Rule.java [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.jmap.api.filtering;
+
+import java.util.Objects;
+
+import com.google.common.base.MoreObjects;
+
+public class Rule {
```
Comment (trantienduchn) [inline server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/Rule.java]: I thought this class is supposed to have more properties than just a single string id

Comment (chibenwa) [reply server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/Rule.java]: I guess it will be enhenced in the near future ;-)

**Comment by trantienduchn** [comment]
I fixed benoi't comments except only one (for debating), fixed commit names then create a new PR because I cannot push my code to anyone's remote brach. See https://github.com/linagora/james-project/pull/1674

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/impl/FilteringAggregate.java (hunk 3 above)
Comment (chibenwa) [reply server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/impl/FilteringAggregate.java]: IMO we should throw on unknown event to not forget to handle new event here...

Comment (mbaechler) [reply server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/impl/FilteringAggregate.java]: not sure about that, what case are you trying to cover this that ?

**Comment by mbaechler** [comment]
closing this one, please follow https://github.com/linagora/james-project/pull/1674 now

**Pull request closed without merging**